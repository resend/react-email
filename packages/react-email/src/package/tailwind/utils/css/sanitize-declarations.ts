import type { AtRule, Root, Rule } from 'postcss';

/**
 * Meant to do all the things necessary, in a per-declaration basis, to have the best email client
 * support possible.
 *
 * Currently it only converts all `rgb` declaration values that use a space-based syntax
 * into a command based one since the space-based isn't well supported.
 */
export const sanitizeDeclarations = (
  nodeContainingDeclarations: Rule | AtRule | Root,
) => {
  nodeContainingDeclarations.walkDecls((declaration) => {
    const rgbParserRegex =
      /rgb\(\s*(\d+)\s*(\d+)\s*(\d+)(?:\s*\/\s*([\d%.]+))?\s*\)/g;

    declaration.value = declaration.value.replaceAll(
      rgbParserRegex,
      (_match, r, g, b, a) => {
        const alpha = a === '1' || typeof a === 'undefined' ? '' : `,${a}`;
        return `rgb(${r},${g},${b}${alpha})`;
      },
    );
  });
};
