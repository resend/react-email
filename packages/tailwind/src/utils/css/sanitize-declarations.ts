import {
  type CssNode,
  type Declaration,
  generate,
  List,
  parse,
  type Value,
  walk,
} from 'css-tree';

const LAB_TO_LMS = {
  l: [0.3963377773761749, 0.2158037573099136],
  m: [-0.1055613458156586, -0.0638541728258133],
  s: [-0.0894841775298119, -1.2914855480194092],
};
const LSM_TO_RGB = {
  r: [4.0767416360759583, -3.3077115392580629, 0.2309699031821043],
  g: [-1.2684379732850315, 2.6097573492876882, -0.341319376002657],
  b: [-0.0041960761386756, -0.7034186179359362, 1.7076146940746117],
};

function lrgbToRgb(input: number) {
  const absoluteNumber = Math.abs(input);
  const sign = input < 0 ? -1 : 1;

  if (absoluteNumber > 0.0031308) {
    return sign * (absoluteNumber ** (1 / 2.4) * 1.055 - 0.055);
  }

  return input * 12.92;
}

function clamp(value: number, min: number, max: number) {
  return Math.min(Math.max(value, min), max);
}

function round(value: number, digits = 1) {
  const factor = 10 ** digits;
  return Math.round(value * factor) / factor;
}

function oklchToOklab(oklch: { l: number; c: number; h: number }) {
  return {
    l: oklch.l,
    a: oklch.c * Math.cos((oklch.h / 180) * Math.PI),
    b: oklch.c * Math.sin((oklch.h / 180) * Math.PI),
  };
}

/** Convert oklab to RGB */
function oklchToRgb(oklch: { l: number; c: number; h: number }) {
  const oklab = oklchToOklab(oklch);

  const l =
    (oklab.l + LAB_TO_LMS.l[0] * oklab.a + LAB_TO_LMS.l[1] * oklab.b) ** 3;
  const m =
    (oklab.l + LAB_TO_LMS.m[0] * oklab.a + LAB_TO_LMS.m[1] * oklab.b) ** 3;
  const s =
    (oklab.l + LAB_TO_LMS.s[0] * oklab.a + LAB_TO_LMS.s[1] * oklab.b) ** 3;

  const r =
    255 *
    lrgbToRgb(LSM_TO_RGB.r[0] * l + LSM_TO_RGB.r[1] * m + LSM_TO_RGB.r[2] * s);
  const g =
    255 *
    lrgbToRgb(LSM_TO_RGB.g[0] * l + LSM_TO_RGB.g[1] * m + LSM_TO_RGB.g[2] * s);
  const b =
    255 *
    lrgbToRgb(LSM_TO_RGB.b[0] * l + LSM_TO_RGB.b[1] * m + LSM_TO_RGB.b[2] * s);

  return {
    r: round(clamp(r, 0, 255)),
    g: round(clamp(g, 0, 255)),
    b: round(clamp(b, 0, 255)),
  };
}

function separteShorthandDeclaration(
  shorthandToReplace: Declaration,
  [start, end]: [string, string],
): Declaration {
  shorthandToReplace.property = start;

  const values =
    shorthandToReplace.value.type === 'Value'
      ? shorthandToReplace.value.children
          .toArray()
          .filter(
            (child) =>
              child.type === 'Dimension' ||
              child.type === 'Number' ||
              child.type === 'Percentage',
          )
      : [shorthandToReplace.value];
  let endValue = shorthandToReplace.value;
  if (values.length === 2) {
    endValue = {
      type: 'Value',
      children: new List<CssNode>().fromArray([values[1]]),
    };
    shorthandToReplace.value = {
      type: 'Value',
      children: new List<CssNode>().fromArray([values[0]]),
    };
  }

  return {
    type: 'Declaration',
    property: end,
    value: endValue,
    important: shorthandToReplace.important,
  };
}

/**
 * Meant to do all the things necessary, in a per-declaration basis, to have the best email client
 * support possible.
 *
 * Here's the transformations it does so far:
 * - convert all `rgb` with space-based syntax into a comma based one;
 * - convert all `oklch` values into `rgb`;
 * - convert all hex values into `rgb`;
 * - convert `padding-inline` into `padding-left` and `padding-right`;
 * - convert `padding-block` into `padding-top` and `padding-bottom`;
 * - convert `margin-inline` into `margin-left` and `margin-right`;
 * - convert `margin-block` into `margin-top` and `margin-bottom`.
 */
export function sanitizeDeclarations(nodeContainingDeclarations: CssNode) {
  walk(nodeContainingDeclarations, {
    visit: 'Declaration',
    enter(declaration, item, list) {
      if (generate(declaration) === 'border-radius:calc(Infinity*1px)') {
        declaration.value = parse('9999px', { context: 'value' }) as Value;
      }
      walk(declaration, {
        visit: 'Function',
        enter(func, funcParentListItem) {
          const children = func.children.toArray();
          if (func.name === 'oklch') {
            let l: number | undefined;
            let c: number | undefined;
            let h: number | undefined;
            let a: number | undefined;
            for (const child of children) {
              if (child.type === 'Number') {
                if (l === undefined) {
                  l = Number.parseFloat(child.value);
                  continue;
                }
                if (c === undefined) {
                  c = Number.parseFloat(child.value);
                  continue;
                }
                if (h === undefined) {
                  h = Number.parseFloat(child.value);
                  continue;
                }
                if (a === undefined) {
                  a = Number.parseFloat(child.value);
                  continue;
                }
              }
              if (child.type === 'Dimension' && child.unit === 'deg') {
                if (h === undefined) {
                  h = Number.parseFloat(child.value);
                  continue;
                }
              }
              if (child.type === 'Percentage') {
                if (l === undefined) {
                  l = Number.parseFloat(child.value) / 100;
                  continue;
                }
                if (a === undefined) {
                  a = Number.parseFloat(child.value) / 100;
                }
              }
            }

            if (l === undefined || c === undefined || h === undefined) {
              throw new Error(
                'Could not determine the parameters of an oklch() function.',
                {
                  cause: declaration,
                },
              );
            }

            const rgb = oklchToRgb({
              l,
              c,
              h,
            });

            const alphaString = a !== undefined ? `,${a}` : '';

            funcParentListItem.data = parse(
              `rgb(${rgb.r},${rgb.g},${rgb.b}${alphaString})`,
              {
                context: 'value',
              },
            );
          }

          if (func.name === 'rgb') {
            let r: number | undefined;
            let g: number | undefined;
            let b: number | undefined;
            let a: number | undefined;
            for (const child of children) {
              if (child.type === 'Number') {
                if (r === undefined) {
                  r = Number.parseFloat(child.value);
                  continue;
                }
                if (g === undefined) {
                  g = Number.parseFloat(child.value);
                  continue;
                }
                if (b === undefined) {
                  b = Number.parseFloat(child.value);
                  continue;
                }
                if (a === undefined) {
                  a = Number.parseFloat(child.value);
                  continue;
                }
              }
              if (child.type === 'Percentage') {
                if (r === undefined) {
                  r = (Number.parseFloat(child.value) * 255) / 100;
                  continue;
                }
                if (g === undefined) {
                  g = (Number.parseFloat(child.value) * 255) / 100;
                  continue;
                }
                if (b === undefined) {
                  b = (Number.parseFloat(child.value) * 255) / 100;
                  continue;
                }
                if (a === undefined) {
                  a = Number.parseFloat(child.value) / 100;
                }
              }
            }

            if (r === undefined || g === undefined || b === undefined) {
              throw new Error(
                'Could not determine the parameters of an rgb() function.',
                {
                  cause: declaration,
                },
              );
            }

            if (a === undefined || a === 1) {
              funcParentListItem.data = parse(`rgb(${r},${g},${b})`, {
                context: 'value',
              });
            } else {
              funcParentListItem.data = parse(`rgb(${r},${g},${b},${a})`, {
                context: 'value',
              });
            }
          }
        },
      });
      walk(declaration, {
        visit: 'Hash',
        enter(hash, hashParentListItem) {
          const hex = hash.value.trim();
          if (hex.length === 3) {
            const r = Number.parseInt(hex.charAt(0) + hex.charAt(0), 16);
            const g = Number.parseInt(hex.charAt(1) + hex.charAt(1), 16);
            const b = Number.parseInt(hex.charAt(2) + hex.charAt(2), 16);
            hashParentListItem.data = parse(`rgb(${r},${g},${b})`, {
              context: 'value',
            });
            return;
          }
          if (hex.length === 4) {
            const r = Number.parseInt(hex.charAt(0) + hex.charAt(0), 16);
            const g = Number.parseInt(hex.charAt(1) + hex.charAt(1), 16);
            const b = Number.parseInt(hex.charAt(2) + hex.charAt(2), 16);
            const a = Number.parseInt(hex.charAt(3) + hex.charAt(3), 16) / 255;
            hashParentListItem.data = parse(
              `rgb(${r},${g},${b},${a.toFixed(1)})`,
              { context: 'value' },
            );
            return;
          }
          if (hex.length === 5) {
            const r = Number.parseInt(hex.slice(0, 2), 16);
            const g = Number.parseInt(hex.charAt(2) + hex.charAt(2), 16);
            const b = Number.parseInt(hex.charAt(3) + hex.charAt(3), 16);
            const a = Number.parseInt(hex.charAt(4) + hex.charAt(4), 16) / 255;
            hashParentListItem.data = parse(
              `rgb(${r},${g},${b},${a.toFixed(1)})`,
              { context: 'value' },
            );
            return;
          }
          if (hex.length === 6) {
            const r = Number.parseInt(hex.slice(0, 2), 16);
            const g = Number.parseInt(hex.slice(2, 4), 16);
            const b = Number.parseInt(hex.slice(4, 6), 16);
            hashParentListItem.data = parse(`rgb(${r},${g},${b})`, {
              context: 'value',
            });
            return;
          }
          if (hex.length === 7) {
            const r = Number.parseInt(hex.slice(0, 2), 16);
            const g = Number.parseInt(hex.slice(2, 4), 16);
            const b = Number.parseInt(hex.slice(4, 6), 16);
            const a = Number.parseInt(hex.charAt(6) + hex.charAt(6), 16) / 255;
            hashParentListItem.data = parse(
              `rgb(${r},${g},${b},${a.toFixed(1)})`,
              { context: 'value' },
            );
            return;
          }
          const r = Number.parseInt(hex.slice(0, 2), 16);
          const g = Number.parseInt(hex.slice(2, 4), 16);
          const b = Number.parseInt(hex.slice(4, 6), 16);
          const a = Number.parseInt(hex.slice(6, 8), 16) / 255;
          hashParentListItem.data = parse(
            `rgb(${r},${g},${b},${a.toFixed(1)})`,
            { context: 'value' },
          );
        },
      });

      if (declaration.property === 'padding-inline') {
        const paddingRight = separteShorthandDeclaration(declaration, [
          'padding-left',
          'padding-right',
        ]);
        list.insertData(paddingRight, item);
      }
      if (declaration.property === 'padding-block') {
        const paddingBottom = separteShorthandDeclaration(declaration, [
          'padding-top',
          'padding-bottom',
        ]);
        list.insertData(paddingBottom, item);
      }
      if (declaration.property === 'margin-inline') {
        const marginRight = separteShorthandDeclaration(declaration, [
          'margin-left',
          'margin-right',
        ]);
        list.insertData(marginRight, item);
      }
      if (declaration.property === 'margin-block') {
        const paddingBottom = separteShorthandDeclaration(declaration, [
          'margin-top',
          'margin-bottom',
        ]);

        list.insertData(paddingBottom, item);
      }
    },
  });
}
