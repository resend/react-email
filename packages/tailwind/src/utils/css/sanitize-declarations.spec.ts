import { generate, parse } from 'css-tree';
import { sanitizeDeclarations } from './sanitize-declarations';

describe('sanitizeDeclarations', () => {
  it('should convert border-radius:calc(Infinity * 1px) to border-radius:9999px', () => {
    const root = parse(`.rounded-full {
  border-radius: calc(Infinity * 1px);
}
`);
    sanitizeDeclarations(root);
    expect(generate(root)).toMatchSnapshot();
  });

  it('should do separation of padding-block and padding-inline', () => {
    let root = parse(`.box {
  padding-inline: 4px 14;
}
`);
    sanitizeDeclarations(root);
    expect(generate(root)).toMatchSnapshot();

    root = parse(`.box {
  padding-block: 10px 20%;
}
`);
    sanitizeDeclarations(root);
    expect(generate(root)).toMatchSnapshot();

    root = parse(`.box {
  padding-inline: 99rem;
}
`);
    sanitizeDeclarations(root);
    expect(generate(root)).toMatchSnapshot();

    root = parse(`.box {
  padding-block: 8px;
}
`);
    sanitizeDeclarations(root);
    expect(generate(root)).toMatchSnapshot();
  });

  it('should do separation of margin-block and margin-inline', () => {
    let root = parse(`.box {
  margin-inline: 4px 14;
}
`);
    sanitizeDeclarations(root);
    expect(generate(root)).toMatchSnapshot();

    root = parse(`.box {
  margin-block: 10px 20%;
}
`);
    sanitizeDeclarations(root);
    expect(generate(root)).toMatchSnapshot();

    root = parse(`.box {
  margin-inline: 99rem;
}
`);
    sanitizeDeclarations(root);
    expect(generate(root)).toMatchSnapshot();

    root = parse(`.box {
  margin-block: 8px;
}
`);
    sanitizeDeclarations(root);
    expect(generate(root)).toMatchSnapshot();
  });

  test('oklch to rgb conversion', () => {
    let stylesheet = parse('div { color: oklch(90.5% 0.2 180); }');
    sanitizeDeclarations(stylesheet);
    expect(generate(stylesheet), 'conversion without alpha').toMatchSnapshot();

    stylesheet = parse('div { color: oklch(96.6% 0.147 107 / 80%); }');
    sanitizeDeclarations(stylesheet);
    expect(generate(stylesheet), 'conversion with alpha').toMatchSnapshot();

    stylesheet = parse('div { color: oklch(96.6% 0.147 107deg / 80%); }');
    sanitizeDeclarations(stylesheet);
    expect(generate(stylesheet), 'conversion with deg unit').toMatchSnapshot();

    stylesheet = parse('div { color: oklch(92.6% 0.0546 218 / 50%); }');
    sanitizeDeclarations(stylesheet);
    expect(generate(stylesheet)).toMatchSnapshot();

    stylesheet = parse('div { color: oklch(88.3% 0.102 329); }');
    sanitizeDeclarations(stylesheet);
    expect(generate(stylesheet)).toMatchSnapshot();

    stylesheet = parse('div { color: oklch(69.3% 0.206 42.8); }');
    sanitizeDeclarations(stylesheet);
    expect(generate(stylesheet)).toMatchSnapshot();
  });

  test('rgba space syntax to comma syntax conversion', () => {
    let stylesheet = parse('div { color: rgb(255 0 128); }');
    sanitizeDeclarations(stylesheet);
    expect(
      generate(stylesheet),
      'conversion with space syntax and no alpha',
    ).toMatchSnapshot();

    stylesheet = parse('div { color: rgb(255 0 128 / 0.5); }');
    sanitizeDeclarations(stylesheet);
    expect(
      generate(stylesheet),
      'conversion with space syntax and alpha',
    ).toMatchSnapshot();

    stylesheet = parse('div { color: rgb(100% 0% 50% / 50%); }');
    sanitizeDeclarations(stylesheet);
    expect(
      generate(stylesheet),
      'conversion with percentage values',
    ).toMatchSnapshot();

    stylesheet = parse('div { color: rgb(255 0 128 / 50%); }');
    sanitizeDeclarations(stylesheet);
    expect(
      generate(stylesheet),
      'conversion with space syntax and alpha with a percetange syntax',
    ).toMatchSnapshot();

    stylesheet = parse('div { color: rgb(255 0 128 / 1); }');
    sanitizeDeclarations(stylesheet);
    expect(generate(stylesheet)).toMatchSnapshot();

    stylesheet = parse(
      'div { background: linear-gradient(rgb(255 0 0), rgb(0 255 0 / 0.8)); }',
    );
    sanitizeDeclarations(stylesheet);
    expect(generate(stylesheet)).toMatchSnapshot();

    stylesheet = parse('div { color: rgb(  255   0   128  /  0.7  ); }');
    sanitizeDeclarations(stylesheet);
    expect(generate(stylesheet)).toMatchSnapshot();

    stylesheet = parse('div { color: rgb(0 0 0); }');
    sanitizeDeclarations(stylesheet);
    expect(generate(stylesheet)).toMatchSnapshot();

    stylesheet = parse('div { color: rgb(255, 0, 128); }');
    sanitizeDeclarations(stylesheet);
    expect(
      generate(stylesheet),
      'treatment for already supported rgb syntax',
    ).toMatchSnapshot();
  });

  test('hex to rgb conversion', () => {
    let stylesheet = parse('div { color: #f0a; }');
    sanitizeDeclarations(stylesheet);
    expect(generate(stylesheet), '3-digit hex without alpha').toMatchSnapshot();

    stylesheet = parse('div { color: #f0a8; }');
    sanitizeDeclarations(stylesheet);
    expect(generate(stylesheet), '4-digit hex with alpha').toMatchSnapshot();

    stylesheet = parse('div { color: #ff00aa; }');
    sanitizeDeclarations(stylesheet);
    expect(generate(stylesheet), '6-digit hex without alpha').toMatchSnapshot();

    stylesheet = parse('div { color: #ff00aa80; }');
    sanitizeDeclarations(stylesheet);
    expect(generate(stylesheet), '8-digit hex with alpha').toMatchSnapshot();

    stylesheet = parse('div { color: #000; }');
    sanitizeDeclarations(stylesheet);
    expect(generate(stylesheet)).toMatchSnapshot();

    stylesheet = parse('div { color: #fff; }');
    sanitizeDeclarations(stylesheet);
    expect(generate(stylesheet)).toMatchSnapshot();

    stylesheet = parse('div { color: #ff0000; }');
    sanitizeDeclarations(stylesheet);
    expect(generate(stylesheet)).toMatchSnapshot();

    stylesheet = parse('div { color: #00ff00; }');
    sanitizeDeclarations(stylesheet);
    expect(generate(stylesheet)).toMatchSnapshot();

    stylesheet = parse('div { color: #0000ff; }');
    sanitizeDeclarations(stylesheet);
    expect(generate(stylesheet)).toMatchSnapshot();

    stylesheet = parse('div { color: #abcdef; }');
    sanitizeDeclarations(stylesheet);
    expect(generate(stylesheet)).toMatchSnapshot();

    stylesheet = parse('div { color: #ABCDEF; }');
    sanitizeDeclarations(stylesheet);
    expect(generate(stylesheet)).toMatchSnapshot();

    stylesheet = parse('div { color: #AbCdEf; }');
    sanitizeDeclarations(stylesheet);
    expect(generate(stylesheet), 'mixed casing').toMatchSnapshot();

    stylesheet = parse('div { color: #ff000000; }');
    sanitizeDeclarations(stylesheet);
    expect(generate(stylesheet)).toMatchSnapshot();

    stylesheet = parse('div { color: #ff0000ff; }');
    sanitizeDeclarations(stylesheet);
    expect(generate(stylesheet)).toMatchSnapshot();

    stylesheet = parse(
      'div { background: linear-gradient(#ff0000, #00ff00, #0000ff); }',
    );
    sanitizeDeclarations(stylesheet);
    expect(generate(stylesheet)).toMatchSnapshot();

    stylesheet = parse(
      'div { background: linear-gradient(#ff0000, rgb(0 255 0), oklch(50% 0.2 240)); }',
    );
    sanitizeDeclarations(stylesheet);
    const result = generate(stylesheet);
    expect(result).toMatchSnapshot();

    stylesheet = parse(
      'div { content: "Visit our site at example.com#section"; }',
    );
    sanitizeDeclarations(stylesheet);
    expect(generate(stylesheet)).toMatchSnapshot();

    stylesheet = parse(`
      div {
        color: #ff0000;
        background-color: #00ff00;
        border-color: #0000ff;
        box-shadow: 0 0 10px #333;
      }
    `);
    sanitizeDeclarations(stylesheet);
    expect(generate(stylesheet)).toMatchSnapshot();
  });

  describe('complex scenarios', () => {
    it('handles multiple declarations with different rgb formats', () => {
      const stylesheet = parse(`
        div {
          color: rgb(255 0 128 / 0.5);
          background-color: rgb(0 255 0);
          border-color: rgb(128, 128, 128);
        }
      `);
      sanitizeDeclarations(stylesheet);
      const result = generate(stylesheet);
      expect(result).toMatchSnapshot();
    });

    it('handles nested rules', () => {
      const stylesheet = parse(`
        @media (min-width: 768px) {
          div { color: rgb(255 0 128 / 0.8); }
        }
      `);
      sanitizeDeclarations(stylesheet);
      expect(generate(stylesheet)).toMatchSnapshot();
    });

    it('processes at-rule declarations', () => {
      const stylesheet = parse(`
        @keyframes fade {
          from { background: rgb(255 0 0 / 0); }
          to { background: rgb(255 0 0 / 1); }
        }
      `);
      sanitizeDeclarations(stylesheet);
      const result = generate(stylesheet);
      expect(result).toMatchSnapshot();
    });
  });
});
