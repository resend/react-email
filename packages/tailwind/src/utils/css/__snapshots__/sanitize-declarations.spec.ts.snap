// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sanitizeDeclarations > complex scenarios > handles multiple declarations with different rgb formats 1`] = `"div{color:rgb(255,0,128,0.5);background-color:rgb(0,255,0);border-color:rgb(128,128,128)}"`;

exports[`sanitizeDeclarations > complex scenarios > handles nested rules 1`] = `"@media (min-width:768px){div{color:rgb(255,0,128,0.8)}}"`;

exports[`sanitizeDeclarations > complex scenarios > processes at-rule declarations 1`] = `"@keyframes fade{from{background:rgb(255,0,0,0)}to{background:rgb(255,0,0)}}"`;

exports[`sanitizeDeclarations > hex to rgb conversion 1`] = `"div{color:rgb(255,0,170)}"`;

exports[`sanitizeDeclarations > hex to rgb conversion 2`] = `"div{color:rgb(255,0,170,0.5)}"`;

exports[`sanitizeDeclarations > hex to rgb conversion 3`] = `"div{color:rgb(255,0,170)}"`;

exports[`sanitizeDeclarations > hex to rgb conversion 4`] = `"div{color:rgb(255,0,170,0.5)}"`;

exports[`sanitizeDeclarations > hex to rgb conversion 5`] = `"div{color:rgb(0,0,0)}"`;

exports[`sanitizeDeclarations > hex to rgb conversion 6`] = `"div{color:rgb(255,255,255)}"`;

exports[`sanitizeDeclarations > hex to rgb conversion 7`] = `"div{color:rgb(255,0,0)}"`;

exports[`sanitizeDeclarations > hex to rgb conversion 8`] = `"div{color:rgb(0,255,0)}"`;

exports[`sanitizeDeclarations > hex to rgb conversion 9`] = `"div{color:rgb(0,0,255)}"`;

exports[`sanitizeDeclarations > hex to rgb conversion 10`] = `"div{color:rgb(171,205,239)}"`;

exports[`sanitizeDeclarations > hex to rgb conversion 11`] = `"div{color:rgb(171,205,239)}"`;

exports[`sanitizeDeclarations > hex to rgb conversion 12`] = `"div{color:rgb(171,205,239)}"`;

exports[`sanitizeDeclarations > hex to rgb conversion 13`] = `"div{color:rgb(255,0,0,0.0)}"`;

exports[`sanitizeDeclarations > hex to rgb conversion 14`] = `"div{color:rgb(255,0,0,1.0)}"`;

exports[`sanitizeDeclarations > hex to rgb conversion 15`] = `"div{background:linear-gradient(rgb(255,0,0),rgb(0,255,0),rgb(0,0,255))}"`;

exports[`sanitizeDeclarations > hex to rgb conversion 16`] = `"div{background:linear-gradient(rgb(255,0,0),rgb(0,255,0),rgb(0,104.6,199))}"`;

exports[`sanitizeDeclarations > hex to rgb conversion 17`] = `"div{content:"Visit our site at example.com#section"}"`;

exports[`sanitizeDeclarations > hex to rgb conversion 18`] = `"div{color:rgb(255,0,0);background-color:rgb(0,255,0);border-color:rgb(0,0,255);box-shadow:0 0 10px rgb(51,51,51)}"`;

exports[`sanitizeDeclarations > oklch to rgb conversion 1`] = `"div{color:rgb(0,255,228.7)}"`;

exports[`sanitizeDeclarations > oklch to rgb conversion 2`] = `"div{color:rgb(255,250.7,125.6,0.8)}"`;

exports[`sanitizeDeclarations > oklch to rgb conversion 3`] = `"div{color:rgb(190.5,240.2,255,0.5)}"`;

exports[`sanitizeDeclarations > oklch to rgb conversion 4`] = `"div{color:rgb(254.6,192.6,249.8)}"`;

exports[`sanitizeDeclarations > oklch to rgb conversion 5`] = `"div{color:rgb(255,99.8,0.1)}"`;

exports[`sanitizeDeclarations > rgba space syntax to comma syntax conversion 1`] = `"div{color:rgb(255,0,128)}"`;

exports[`sanitizeDeclarations > rgba space syntax to comma syntax conversion 2`] = `"div{color:rgb(255,0,128,0.5)}"`;

exports[`sanitizeDeclarations > rgba space syntax to comma syntax conversion 3`] = `"div{color:rgb(255,0,128,0.5)}"`;

exports[`sanitizeDeclarations > rgba space syntax to comma syntax conversion 4`] = `"div{color:rgb(255,0,128)}"`;

exports[`sanitizeDeclarations > rgba space syntax to comma syntax conversion 5`] = `"div{background:linear-gradient(rgb(255,0,0),rgb(0,255,0,0.8))}"`;

exports[`sanitizeDeclarations > rgba space syntax to comma syntax conversion 6`] = `"div{color:rgb(255,0,128,0.7)}"`;

exports[`sanitizeDeclarations > rgba space syntax to comma syntax conversion 7`] = `"div{color:rgb(0,0,0)}"`;

exports[`sanitizeDeclarations > rgba space syntax to comma syntax conversion 8`] = `"div{color:rgb(255,0,128)}"`;
